abci/multiplexer.go

```diff
@@ -4,36 +4,72 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/01builders/nova/internal"
+	cmtcfg "github.com/cometbft/cometbft/config"
+	"github.com/cometbft/cometbft/node"
+	"github.com/cometbft/cometbft/p2p"
+	pvm "github.com/cometbft/cometbft/privval"
+	"github.com/cometbft/cometbft/proxy"
+	"github.com/cometbft/cometbft/rpc/client/local"
+	db "github.com/cosmos/cosmos-db"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/server"
+	"github.com/cosmos/cosmos-sdk/server/api"
+	serverconfig "github.com/cosmos/cosmos-sdk/server/config"
+	servergrpc "github.com/cosmos/cosmos-sdk/server/grpc"
+	servercmtlog "github.com/cosmos/cosmos-sdk/server/log"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	"github.com/cosmos/cosmos-sdk/version"
+	"github.com/hashicorp/go-metrics"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/credentials/insecure"
+	"io"
 	"math"
+	"net"
 	"os"
+	"os/signal"
+	"path/filepath"
 	"strings"
 	"sync"
-
-	"github.com/spf13/viper"
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials/insecure"
+	"syscall"
 
 	"cosmossdk.io/log"
-	abci "github.com/cometbft/cometbft/abci/types"
-	"github.com/cometbft/cometbft/proxy"
 	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+	"google.golang.org/grpc"
 
 	"github.com/01builders/nova/appd"
 )
 
+const (
+	flagTraceStore = "trace-store"
+	flagGRPCOnly   = "grpc-only"
+)
+
 type Multiplexer struct {
 	logger log.Logger
 	mu     sync.Mutex
 
+	svrCtx        *server.Context
+	svrCfg        serverconfig.Config
+	cmtCfg        *cmtcfg.Config
+	clientContext client.Context
+
 	lastAppVersion uint64
-	started        bool
 
-	latestApp     servertypes.ABCI
+	started       bool
+	appCreator    servertypes.AppCreator
+	latestApp     servertypes.Application
 	activeVersion Version
 	chainID       string
 
-	versions Versions
-	conn     *grpc.ClientConn
+	tmNode     *node.Node
+	versions   Versions
+	conn       *grpc.ClientConn
+	cleanupFns []func() error
+
+	ctx context.Context
+	g   *errgroup.Group
```

Comment:
this feels a little hacky, and am not 100% sure if it will work, I will explain more about it where it is used.